fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/InSource.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * InSource.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2021 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.query.plan.plans;
+
+import com.apple.foundationdb.record.EvaluationContext;
+import com.apple.foundationdb.record.ObjectPlanHash;
+import com.apple.foundationdb.record.PlanHashable;
+import com.apple.foundationdb.record.query.plan.temp.Quantifier;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+
+public abstract class InSource implements PlanHashable {
+    @SuppressWarnings({"unchecked", "rawtypes"})
+    private static final Comparator<Object> VALUE_COMPARATOR = Comparator.comparing(o -> ((Comparable)o));
+
+    @Nonnull
+    private final String bindingName;
+
+    protected InSource(@Nonnull final String bindingName) {
+        this.bindingName = bindingName;
+    }
+
+    @Nonnull
+    public String getBindingName() {
+        return bindingName;
+    }
+
+    public abstract boolean isSorted();
+
+    public abstract boolean isReverse();
+
+    protected abstract int size(@Nonnull EvaluationContext context);
+
+    @Nonnull
+    public List<Object> getValues() {
+        return getValues(null);
+    }
+
+    @Nonnull
+    protected abstract List<Object> getValues(@Nullable EvaluationContext context);
+
+    @Nonnull
+    public abstract RecordQueryInJoinPlan toInJoinPlan(@Nonnull final Quantifier.Physical innerQuantifier);
+
+    public int baseHash(@Nonnull final PlanHashKind hashKind, @Nonnull ObjectPlanHash objectPlanHash) {
+        return objectPlanHash.planHash(hashKind);
+    }
+
+    @Nonnull
+    public static List<Object> sortValues(@Nonnull List<Object> values, final boolean isReversed) {
+        if (values.size() < 2 ) {
+            return values;
```

Comment:
I think this is fine and also just coming from a refactor. The function's contract is to just return the sorted list back without changing the original input list. The copy happens since `.sort()` is sorting in place.

In order to make the contract stronger I will wrap the singleton element into aniqmmutable list factory method.

Reply:
Sounds good.